// @tiendanube/nube-sdk-jsx/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Accordion: () => Accordion,
  Box: () => Box,
  Button: () => Button,
  Check: () => Check,
  Checkbox: () => Checkbox,
  Col: () => Col,
  Column: () => Column,
  Field: () => Field,
  Fragment: () => Fragment,
  Icon: () => Icon,
  Iframe: () => Iframe,
  Image: () => Image,
  Img: () => Img,
  Link: () => Link,
  NumberField: () => NumberField,
  Progress: () => Progress,
  Row: () => Row,
  Select: () => Select,
  Svg: () => Svg,
  Text: () => Text,
  Textarea: () => Textarea,
  Toast: () => Toast,
  Txt: () => Txt,
  Txtarea: () => Txtarea
});
module.exports = __toCommonJS(index_exports);

// src/components.tsx
var import_nube_sdk_ui = require("@tiendanube/nube-sdk-ui");
function Box(props) {
  return (0, import_nube_sdk_ui.box)(props);
}
function Col(props) {
  return (0, import_nube_sdk_ui.column)(props);
}
function Column(props) {
  return (0, import_nube_sdk_ui.column)(props);
}
function Row(props) {
  return (0, import_nube_sdk_ui.row)(props);
}
function Field(props) {
  return (0, import_nube_sdk_ui.field)(props);
}
function NumberField(props) {
  return (0, import_nube_sdk_ui.numberfield)(props);
}
function Button(props) {
  return (0, import_nube_sdk_ui.button)(props);
}
function Link(props) {
  return (0, import_nube_sdk_ui.link)(props);
}
function Fragment(props) {
  return (0, import_nube_sdk_ui.fragment)(props);
}
function Txt(props) {
  return (0, import_nube_sdk_ui.text)(props);
}
function Text(props) {
  return (0, import_nube_sdk_ui.text)(props);
}
function Check(props) {
  return (0, import_nube_sdk_ui.checkbox)(props);
}
function Checkbox(props) {
  return (0, import_nube_sdk_ui.checkbox)(props);
}
function Txtarea(props) {
  return (0, import_nube_sdk_ui.textarea)(props);
}
function Textarea(props) {
  return (0, import_nube_sdk_ui.textarea)(props);
}
function Img(props) {
  return (0, import_nube_sdk_ui.image)(props);
}
function Image(props) {
  return (0, import_nube_sdk_ui.image)(props);
}
function Progress(props) {
  return (0, import_nube_sdk_ui.progress)(props);
}
function Iframe(props) {
  return (0, import_nube_sdk_ui.iframe)(props);
}
function Select(props) {
  return (0, import_nube_sdk_ui.select)(props);
}
function AccordionRoot(props) {
  return (0, import_nube_sdk_ui.accordionRoot)(props);
}
function AccordionItem(props) {
  return (0, import_nube_sdk_ui.accordionItem)(props);
}
function AccordionContent(props) {
  return (0, import_nube_sdk_ui.accordionContent)(props);
}
function AccordionHeader(props) {
  return (0, import_nube_sdk_ui.accordionHeader)(props);
}
var Accordion = {
  Root: AccordionRoot,
  Item: AccordionItem,
  Content: AccordionContent,
  Header: AccordionHeader
};
function ToastRoot(props) {
  return (0, import_nube_sdk_ui.toastRoot)(props);
}
function ToastTitle(props) {
  return (0, import_nube_sdk_ui.toastTitle)(props);
}
function ToastDescription(props) {
  return (0, import_nube_sdk_ui.toastDescription)(props);
}
var Toast = {
  Root: ToastRoot,
  Title: ToastTitle,
  Description: ToastDescription
};
function Icon(props) {
  return (0, import_nube_sdk_ui.icon)(props);
}
function SvgRoot(props) {
  return (0, import_nube_sdk_ui.svgRoot)(props);
}
function SvgCircle(props) {
  return (0, import_nube_sdk_ui.svgCircle)(props);
}
function SvgPath(props) {
  return (0, import_nube_sdk_ui.svgPath)(props);
}
function SvgG(props) {
  return (0, import_nube_sdk_ui.svgG)(props);
}
function SvgRect(props) {
  return (0, import_nube_sdk_ui.svgRect)(props);
}
function SvgLine(props) {
  return (0, import_nube_sdk_ui.svgLine)(props);
}
function SvgEllipse(props) {
  return (0, import_nube_sdk_ui.svgEllipse)(props);
}
function SvgPolygon(props) {
  return (0, import_nube_sdk_ui.svgPolygon)(props);
}
function SvgPolyline(props) {
  return (0, import_nube_sdk_ui.svgPolyline)(props);
}
function SvgText(props) {
  return (0, import_nube_sdk_ui.svgText)(props);
}
function SvgTSpan(props) {
  return (0, import_nube_sdk_ui.svgTspan)(props);
}
function SvgDefs(props) {
  return (0, import_nube_sdk_ui.svgDefs)(props);
}
function SvgStop(props) {
  return (0, import_nube_sdk_ui.svgStop)(props);
}
function SvgLinearGradient(props) {
  return (0, import_nube_sdk_ui.svgLinearGradient)(props);
}
function SvgRadialGradient(props) {
  return (0, import_nube_sdk_ui.svgRadialGradient)(props);
}
function SvgMask(props) {
  return (0, import_nube_sdk_ui.svgMask)(props);
}
function SvgClipPath(props) {
  return (0, import_nube_sdk_ui.svgClipPath)(props);
}
function SvgUse(props) {
  return (0, import_nube_sdk_ui.svgUse)(props);
}
function SvgSymbol(props) {
  return (0, import_nube_sdk_ui.svgSymbol)(props);
}
function SvgPattern(props) {
  return (0, import_nube_sdk_ui.svgPattern)(props);
}
function SvgFilter(props) {
  return (0, import_nube_sdk_ui.svgFilter)(props);
}
function SvgFeGaussianBlur(props) {
  return (0, import_nube_sdk_ui.svgFeGaussianBlur)(props);
}
function SvgFeOffset(props) {
  return (0, import_nube_sdk_ui.svgFeOffset)(props);
}
function SvgFeMerge(props) {
  return (0, import_nube_sdk_ui.svgFeMerge)(props);
}
function SvgFeMergeNode(props) {
  return (0, import_nube_sdk_ui.svgFeMergeNode)(props);
}
var Svg = {
  Root: SvgRoot,
  Circle: SvgCircle,
  Path: SvgPath,
  G: SvgG,
  Rect: SvgRect,
  Line: SvgLine,
  Ellipse: SvgEllipse,
  Polygon: SvgPolygon,
  Polyline: SvgPolyline,
  Text: SvgText,
  TSpan: SvgTSpan,
  Defs: SvgDefs,
  Stop: SvgStop,
  LinearGradient: SvgLinearGradient,
  RadialGradient: SvgRadialGradient,
  Mask: SvgMask,
  ClipPath: SvgClipPath,
  Use: SvgUse,
  Symbol: SvgSymbol,
  Pattern: SvgPattern,
  Filter: SvgFilter,
  FeGaussianBlur: SvgFeGaussianBlur,
  FeOffset: SvgFeOffset,
  FeMerge: SvgFeMerge,
  FeMergeNode: SvgFeMergeNode
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Accordion,
  Box,
  Button,
  Check,
  Checkbox,
  Col,
  Column,
  Field,
  Fragment,
  Icon,
  Iframe,
  Image,
  Img,
  Link,
  NumberField,
  Progress,
  Row,
  Select,
  Svg,
  Text,
  Textarea,
  Toast,
  Txt,
  Txtarea
});
