// @tiendanube/nube-sdk-jsx/dist/index.js
// src/components.tsx
import {
  accordionContent,
  accordionHeader,
  accordionItem,
  accordionRoot,
  box,
  button,
  checkbox,
  column,
  field,
  fragment,
  icon,
  iframe,
  image,
  link,
  numberfield,
  progress,
  row,
  select,
  svgCircle,
  svgClipPath,
  svgDefs,
  svgEllipse,
  svgFeGaussianBlur,
  svgFeMerge,
  svgFeMergeNode,
  svgFeOffset,
  svgFilter,
  svgG,
  svgLine,
  svgLinearGradient,
  svgMask,
  svgPath,
  svgPattern,
  svgPolygon,
  svgPolyline,
  svgRadialGradient,
  svgRect,
  svgRoot,
  svgStop,
  svgSymbol,
  svgText,
  svgTspan,
  svgUse,
  text,
  textarea,
  toastDescription,
  toastRoot,
  toastTitle
} from "@tiendanube/nube-sdk-ui";
function Box(props) {
  return box(props);
}
function Col(props) {
  return column(props);
}
function Column(props) {
  return column(props);
}
function Row(props) {
  return row(props);
}
function Field(props) {
  return field(props);
}
function NumberField(props) {
  return numberfield(props);
}
function Button(props) {
  return button(props);
}
function Link(props) {
  return link(props);
}
function Fragment(props) {
  return fragment(props);
}
function Txt(props) {
  return text(props);
}
function Text(props) {
  return text(props);
}
function Check(props) {
  return checkbox(props);
}
function Checkbox(props) {
  return checkbox(props);
}
function Txtarea(props) {
  return textarea(props);
}
function Textarea(props) {
  return textarea(props);
}
function Img(props) {
  return image(props);
}
function Image(props) {
  return image(props);
}
function Progress(props) {
  return progress(props);
}
function Iframe(props) {
  return iframe(props);
}
function Select(props) {
  return select(props);
}
function AccordionRoot(props) {
  return accordionRoot(props);
}
function AccordionItem(props) {
  return accordionItem(props);
}
function AccordionContent(props) {
  return accordionContent(props);
}
function AccordionHeader(props) {
  return accordionHeader(props);
}
var Accordion = {
  Root: AccordionRoot,
  Item: AccordionItem,
  Content: AccordionContent,
  Header: AccordionHeader
};
function ToastRoot(props) {
  return toastRoot(props);
}
function ToastTitle(props) {
  return toastTitle(props);
}
function ToastDescription(props) {
  return toastDescription(props);
}
var Toast = {
  Root: ToastRoot,
  Title: ToastTitle,
  Description: ToastDescription
};
function Icon(props) {
  return icon(props);
}
function SvgRoot(props) {
  return svgRoot(props);
}
function SvgCircle(props) {
  return svgCircle(props);
}
function SvgPath(props) {
  return svgPath(props);
}
function SvgG(props) {
  return svgG(props);
}
function SvgRect(props) {
  return svgRect(props);
}
function SvgLine(props) {
  return svgLine(props);
}
function SvgEllipse(props) {
  return svgEllipse(props);
}
function SvgPolygon(props) {
  return svgPolygon(props);
}
function SvgPolyline(props) {
  return svgPolyline(props);
}
function SvgText(props) {
  return svgText(props);
}
function SvgTSpan(props) {
  return svgTspan(props);
}
function SvgDefs(props) {
  return svgDefs(props);
}
function SvgStop(props) {
  return svgStop(props);
}
function SvgLinearGradient(props) {
  return svgLinearGradient(props);
}
function SvgRadialGradient(props) {
  return svgRadialGradient(props);
}
function SvgMask(props) {
  return svgMask(props);
}
function SvgClipPath(props) {
  return svgClipPath(props);
}
function SvgUse(props) {
  return svgUse(props);
}
function SvgSymbol(props) {
  return svgSymbol(props);
}
function SvgPattern(props) {
  return svgPattern(props);
}
function SvgFilter(props) {
  return svgFilter(props);
}
function SvgFeGaussianBlur(props) {
  return svgFeGaussianBlur(props);
}
function SvgFeOffset(props) {
  return svgFeOffset(props);
}
function SvgFeMerge(props) {
  return svgFeMerge(props);
}
function SvgFeMergeNode(props) {
  return svgFeMergeNode(props);
}
var Svg = {
  Root: SvgRoot,
  Circle: SvgCircle,
  Path: SvgPath,
  G: SvgG,
  Rect: SvgRect,
  Line: SvgLine,
  Ellipse: SvgEllipse,
  Polygon: SvgPolygon,
  Polyline: SvgPolyline,
  Text: SvgText,
  TSpan: SvgTSpan,
  Defs: SvgDefs,
  Stop: SvgStop,
  LinearGradient: SvgLinearGradient,
  RadialGradient: SvgRadialGradient,
  Mask: SvgMask,
  ClipPath: SvgClipPath,
  Use: SvgUse,
  Symbol: SvgSymbol,
  Pattern: SvgPattern,
  Filter: SvgFilter,
  FeGaussianBlur: SvgFeGaussianBlur,
  FeOffset: SvgFeOffset,
  FeMerge: SvgFeMerge,
  FeMergeNode: SvgFeMergeNode
};
export {
  Accordion,
  Box,
  Button,
  Check,
  Checkbox,
  Col,
  Column,
  Field,
  Fragment,
  Icon,
  Iframe,
  Image,
  Img,
  Link,
  NumberField,
  Progress,
  Row,
  Select,
  Svg,
  Text,
  Textarea,
  Toast,
  Txt,
  Txtarea
};
