// @tiendanube/nube-sdk-ui/dist/index.js
// src/components/generateInternalId.ts
var memo = /* @__PURE__ */ new WeakMap();
function stableStringify(value) {
  return JSON.stringify(value, (key, val) => {
    if (key === "children" && typeof val !== "string") return "[children]";
    if (typeof val === "function") {
      if (memo.has(val)) return memo.get(val);
      const fnid = `[fn:${crypto.randomUUID()}]`;
      memo.set(val, fnid);
      return fnid;
    }
    return val;
  });
}
function djb2Hash(str) {
  let hash = 5381;
  for (let i = 0; i < str.length; i++) {
    hash = hash * 33 ^ str.charCodeAt(i);
  }
  return (hash >>> 0).toString(36);
}
function createId(type, hash) {
  return `${type}-${self.__APP_DATA__.id}-${hash}`;
}
function createInternalIdRegex(appId) {
  const escapedAppId = appId.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`^[a-zA-Z0-9_-]+-${escapedAppId}-[a-z0-9]+$`);
}
function isValidInternalId(id) {
  if (typeof id !== "string" || id === "") return false;
  const regex = createInternalIdRegex(self.__APP_DATA__.id);
  return regex.test(id);
}
function generateInternalId(type, props) {
  if (isValidInternalId(props.__internalId)) return props.__internalId;
  const cached = memo.get(props);
  if (cached) return createId(type, cached);
  const key = stableStringify(props);
  const hash = djb2Hash(key);
  memo.set(props, hash);
  return createId(type, hash);
}

// src/components/box.ts
var box = (props) => ({
  type: "box",
  ...props,
  __internalId: generateInternalId("box", props)
});

// src/components/column.ts
var column = (props) => ({
  type: "col",
  ...props,
  __internalId: generateInternalId("col", props)
});

// src/components/row.ts
var row = (props) => ({
  type: "row",
  ...props,
  __internalId: generateInternalId("row", props)
});

// src/components/field.ts
var field = (props) => ({
  type: "field",
  ...props,
  __internalId: generateInternalId("field", props)
});

// src/components/number-field.ts
var numberfield = (props) => ({
  type: "numberfield",
  ...props,
  __internalId: generateInternalId("numberfield", props)
});

// src/components/fragment.ts
var fragment = (props) => ({
  type: "fragment",
  ...props,
  __internalId: generateInternalId("fragment", props)
});

// src/components/image.ts
var img = (props) => ({
  type: "img",
  ...props,
  __internalId: generateInternalId("img", props)
});
var image = (props) => ({
  type: "img",
  ...props,
  __internalId: generateInternalId("img", props)
});

// src/components/progress.ts
var progress = (props) => ({
  type: "progress",
  ...props,
  __internalId: generateInternalId("progress", props)
});

// src/components/iframe.ts
var iframe = (props) => ({
  type: "iframe",
  ...props,
  __internalId: generateInternalId("iframe", props)
});

// src/components/text.ts
var text = (props) => ({
  type: "txt",
  ...props,
  __internalId: generateInternalId("txt", props)
});
var txt = (props) => ({
  type: "txt",
  ...props,
  __internalId: generateInternalId("txt", props)
});

// src/components/checkbox.ts
var check = (props) => ({
  type: "check",
  ...props,
  __internalId: generateInternalId("check", props)
});
var checkbox = (props) => ({
  type: "check",
  ...props
});

// src/components/textarea.ts
var txtarea = (props) => ({
  type: "txtarea",
  ...props,
  __internalId: generateInternalId("txtarea", props)
});
var textarea = (props) => ({
  type: "txtarea",
  ...props,
  __internalId: generateInternalId("txtarea", props)
});

// src/components/button.ts
var button = (props) => ({
  type: "button",
  ...props,
  __internalId: generateInternalId("button", props)
});

// src/components/select.ts
var select = (props) => ({
  type: "select",
  ...props,
  __internalId: generateInternalId("select", props)
});

// src/components/link.ts
var link = (props) => ({
  type: "link",
  ...props,
  __internalId: generateInternalId("link", props)
});

// src/components/accordion.ts
var accordionRoot = (props) => ({
  type: "accordionRoot",
  ...props,
  __internalId: generateInternalId("accordionRoot", props)
});

// src/components/accordion-item.ts
var accordionItem = (props) => ({
  type: "accordionItem",
  ...props,
  __internalId: generateInternalId("accordionItem", props)
});

// src/components/accordion-content.ts
var accordionContent = (props) => ({
  type: "accordionContent",
  ...props,
  __internalId: generateInternalId("accordionContent", props)
});

// src/components/accordion.header.ts
var accordionHeader = (props) => ({
  type: "accordionHeader",
  ...props,
  __internalId: generateInternalId("accordionHeader", props)
});

// src/components/toast-root.ts
var toastRoot = (props) => ({
  type: "toastRoot",
  ...props,
  __internalId: generateInternalId("toastRoot", props)
});

// src/components/toast-title.ts
var toastTitle = (props) => ({
  type: "toastTitle",
  ...props,
  __internalId: generateInternalId("toastTitle", props)
});

// src/components/toast-description.ts
var toastDescription = (props) => ({
  type: "toastDescription",
  ...props,
  __internalId: generateInternalId("toastDescription", props)
});

// src/components/icon.ts
var icon = (props) => ({
  type: "icon",
  ...props,
  __internalId: generateInternalId("icon", props)
});

// src/components/svg-root.ts
var svgRoot = (props) => ({
  type: "svgRoot",
  ...props,
  __internalId: generateInternalId("svgRoot", props)
});

// src/components/svg-circle.ts
var svgCircle = (props) => ({
  type: "svgCircle",
  ...props,
  __internalId: generateInternalId("svgCircle", props)
});

// src/components/svg-path.ts
var svgPath = (props) => ({
  type: "svgPath",
  ...props,
  __internalId: generateInternalId("svgPath", props)
});

// src/components/svg-g.ts
var svgG = (props) => ({
  type: "svgG",
  ...props,
  __internalId: generateInternalId("svgG", props)
});

// src/components/svg-rect.ts
var svgRect = (props) => ({
  type: "svgRect",
  ...props,
  __internalId: generateInternalId("svgRect", props)
});

// src/components/svg-line.ts
var svgLine = (props) => ({
  type: "svgLine",
  ...props,
  __internalId: generateInternalId("svgLine", props)
});

// src/components/svg-ellipse.ts
var svgEllipse = (props) => ({
  type: "svgEllipse",
  ...props,
  __internalId: generateInternalId("svgEllipse", props)
});

// src/components/svg-polygon.ts
var svgPolygon = (props) => ({
  type: "svgPolygon",
  ...props,
  __internalId: generateInternalId("svgPolygon", props)
});

// src/components/svg-polyline.ts
var svgPolyline = (props) => ({
  type: "svgPolyline",
  ...props,
  __internalId: generateInternalId("svgPolyline", props)
});

// src/components/svg-text.ts
var svgText = (props) => ({
  type: "svgText",
  ...props,
  __internalId: generateInternalId("svgText", props)
});

// src/components/svg-tspan.ts
var svgTspan = (props) => ({
  type: "svgTspan",
  ...props,
  __internalId: generateInternalId("svgTspan", props)
});

// src/components/svg-defs.ts
var svgDefs = (props) => ({
  type: "svgDefs",
  ...props,
  __internalId: generateInternalId("svgDefs", props)
});

// src/components/svg-stop.ts
var svgStop = (props) => ({
  type: "svgStop",
  ...props,
  __internalId: generateInternalId("svgStop", props)
});

// src/components/svg-linear-gradient.ts
var svgLinearGradient = (props) => ({
  type: "svgLinearGradient",
  ...props,
  __internalId: generateInternalId("svgLinearGradient", props)
});

// src/components/svg-radial-gradient.ts
var svgRadialGradient = (props) => ({
  type: "svgRadialGradient",
  ...props,
  __internalId: generateInternalId("svgRadialGradient", props)
});

// src/components/svg-mask.ts
var svgMask = (props) => ({
  type: "svgMask",
  ...props,
  __internalId: generateInternalId("svgMask", props)
});

// src/components/svg-clip-path.ts
var svgClipPath = (props) => ({
  type: "svgClipPath",
  ...props,
  __internalId: generateInternalId("svgClipPath", props)
});

// src/components/svg-use.ts
var svgUse = (props) => ({
  type: "svgUse",
  ...props,
  __internalId: generateInternalId("svgUse", props)
});

// src/components/svg-symbol.ts
var svgSymbol = (props) => ({
  type: "svgSymbol",
  ...props,
  __internalId: generateInternalId("svgSymbol", props)
});

// src/components/svg-pattern.ts
var svgPattern = (props) => ({
  type: "svgPattern",
  ...props,
  __internalId: generateInternalId("svgPattern", props)
});

// src/components/svg-filter.ts
var svgFilter = (props) => ({
  type: "svgFilter",
  ...props,
  __internalId: generateInternalId("svgFilter", props)
});

// src/components/svg-fe-gaussian-blur.ts
var svgFeGaussianBlur = (props) => ({
  type: "svgFeGaussianBlur",
  ...props,
  __internalId: generateInternalId("svgFeGaussianBlur", props)
});

// src/components/svg-fe-offset.ts
var svgFeOffset = (props) => ({
  type: "svgFeOffset",
  ...props,
  __internalId: generateInternalId("svgFeOffset", props)
});

// src/components/svg-fe-merge.ts
var svgFeMerge = (props) => ({
  type: "svgFeMerge",
  ...props,
  __internalId: generateInternalId("svgFeMerge", props)
});

// src/components/svg-fe-merge-node.ts
var svgFeMergeNode = (props) => ({
  type: "svgFeMergeNode",
  ...props,
  __internalId: generateInternalId("svgFeMergeNode", props)
});

// src/styles/ThemeColor.ts
var ThemeColor = class {
  constructor(token) {
    this.token = token;
  }
  opacity(opacity) {
    return `var(--${this.token}-opacity-${opacity.toString().padStart(2, "0")})`;
  }
  toValue() {
    return `var(--${this.token})`;
  }
  toString() {
    return this.toValue();
  }
};

// src/styles/StyleSheet.ts
function parseValue(value) {
  if (value instanceof ThemeColor) {
    return value.toValue();
  }
  return value;
}
function parseStyleObject(style) {
  const parsed = {};
  for (const key in style) {
    const val = style[key];
    if (val && typeof val === "object" && !Array.isArray(val) && !(val instanceof ThemeColor) && Object.prototype.hasOwnProperty.call(val, "toString")) {
      parsed[key] = parseStyleObject(val);
    } else {
      parsed[key] = parseValue(val);
    }
  }
  return parsed;
}
var StyleSheet = {
  create(styles) {
    const parsedStyles = {};
    for (const key in styles) {
      parsedStyles[key] = parseStyleObject(
        styles[key]
      );
    }
    return parsedStyles;
  }
};

// src/styles/theme.ts
var theme = {
  color: {
    accent: new ThemeColor("accent-color"),
    main: {
      foreground: new ThemeColor("main-foreground"),
      background: new ThemeColor("main-background")
    }
  },
  border: {
    color: "var(--border-color)",
    radius: "var(--border-radius)"
  },
  box: {
    border: {
      color: "var(--box-border-color)",
      radius: "var(--box-border-radius)"
    }
  },
  input: {
    border: {
      color: "var(--input-border-color)"
    }
  },
  button: {
    foreground: "var(--button-foreground)",
    background: "var(--button-background)",
    borderColor: "var(--button-border-color)",
    borderRadius: "var(--button-border-radius)"
  },
  label: {
    foreground: "var(--label-foreground)",
    background: "var(--label-background)"
  },
  header: {
    foreground: "var(--header-foreground)",
    background: "var(--header-background)",
    logo: {
      maxWidth: "var(--header-logo-max-width)",
      maxHeight: "var(--header-logo-max-height)",
      font: "var(--header-logo-font)",
      fontSize: "var(--header-logo-font-size)",
      fontWeight: "var(--header-logo-font-weight)",
      textTransform: "var(--header-logo-text-transform)",
      letterSpacing: "var(--header-logo-letter-spacing)"
    }
  },
  footer: {
    foreground: "var(--footer-foreground)",
    background: "var(--footer-background)"
  },
  heading: {
    font: "var(--heading-font)",
    fontWeight: "var(--heading-font-weight)",
    textTransform: "var(--heading-text-transform)",
    letterSpacing: "var(--heading-letter-spacing)"
  }
};

// src/styles/minify.ts
function minify(css) {
  return css.replace(/\/\*[\s\S]*?\*\//g, "").replace(/\s*([\{\}:;,])\s*/g, "$1").replace(/\s+/g, " ").replace(/;\}/g, "}").replace(/\n/g, "").trim();
}

// src/styles/shortid.ts
function shortid(length = 8) {
  const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  const array = new Uint8Array(length);
  crypto.getRandomValues(array);
  return Array.from(array, (byte) => chars[byte % chars.length]).join("");
}

// src/styles/keyframes.ts
function keyframes(strings, ...exprs) {
  const rawCSS = String.raw({ raw: strings }, ...exprs);
  const css = minify(rawCSS);
  const id = `kf-${shortid()}`;
  return {
    name: id,
    css: `@keyframes ${id}{${css}}`,
    toString() {
      return id;
    }
  };
}
function isKeyframesObject(value) {
  return typeof value === "object" && value !== null && "css" in value && "name" in value && "toString" in value && typeof value.toString === "function";
}

// src/styles/styled.ts
function styled(baseComponent) {
  return (strings, ...exprs) => {
    const StyledComponent = (props) => {
      let rawCSS = "";
      const animation = [];
      for (let i = 0; i < strings.length; i++) {
        const currentString = strings[i] ?? "";
        const expr = exprs[i];
        if (isKeyframesObject(expr)) {
          animation.push(expr.css);
          rawCSS += `${currentString}${expr.name} `;
        } else {
          const value = expr === null || expr === void 0 ? "" : String(expr);
          rawCSS += `${currentString}${value}`;
        }
      }
      const minifiedCSS = minify(`${animation.join("")}${rawCSS}`);
      const componentType = baseComponent.name || "component";
      const internalId = generateInternalId(componentType, {
        ...props,
        __styledCSS: minifiedCSS
      });
      const component = baseComponent({ ...props, __internalId: internalId });
      if (typeof component === "string") return component;
      component.styled = `${component.styled || ""}${minifiedCSS}`;
      return component;
    };
    return StyledComponent;
  };
}
export {
  StyleSheet,
  accordionContent,
  accordionHeader,
  accordionItem,
  accordionRoot,
  box,
  button,
  check,
  checkbox,
  column,
  field,
  fragment,
  icon,
  iframe,
  image,
  img,
  keyframes,
  link,
  numberfield,
  progress,
  row,
  select,
  styled,
  svgCircle,
  svgClipPath,
  svgDefs,
  svgEllipse,
  svgFeGaussianBlur,
  svgFeMerge,
  svgFeMergeNode,
  svgFeOffset,
  svgFilter,
  svgG,
  svgLine,
  svgLinearGradient,
  svgMask,
  svgPath,
  svgPattern,
  svgPolygon,
  svgPolyline,
  svgRadialGradient,
  svgRect,
  svgRoot,
  svgStop,
  svgSymbol,
  svgText,
  svgTspan,
  svgUse,
  text,
  textarea,
  theme,
  toastDescription,
  toastRoot,
  toastTitle,
  txt,
  txtarea
};
